Initial stuff for ansible-database-migrations

moshthepitt opened the pull request:
This PR adds the `ansible-database-migrations` role.

The purpose of this role is to build on top of [ansible-sqitch](https://github.com/onaio/ansible-sqitch) and actually run sqitch migrations.  Actually, that is not correct.  This role does not just run sqitch migrations, it does three things:

1. Runs any pre-defined SQL queries that are meant to set up the database.  These may be things such as creating databases and database users, creating postgres extensions etc.
2. Actually running sqitch migrations
3. Setting up certain pre-defined SQL queries to be run periodically by setting them up as cron jobs.  These may be queries such as a daily data dump, refreshing materialized views, etc.

### Other things

- Currently only works for postgres
- Sets up and uses a .pgpass file for authentication
- Periodic SQL query results are output to predefined log files, and logrotate configs are set for these

### Things this role does not do

- This role does not concern itself with installing the database system
- It does not install sqitch directly, that is done through  [ansible-sqitch](https://github.com/onaio/ansible-sqitch)

moshthepitt committed:
Add python requirements

moshthepitt committed:
Add .venv to gitignore

moshthepitt committed:
Add ansible requirements

moshthepitt committed:
Add git clone task

This task simply clones repos that contain database migrations

moshthepitt committed:
Add initial default vars

moshthepitt committed:
Update meta

- Add correct role information
- Add sqitch role dependency

moshthepitt committed:
Ensure molecule tests work

- Add correct role information
- Enable ssh agent forwarding

moshthepitt committed:
Fix lint isues

moshthepitt committed:
Move variables to converge playbook

moshthepitt committed:
Add mechanism to copy DB credentials to pgpass

moshthepitt committed:
Fix typo in vars

moshthepitt committed:
Clean up code

moshthepitt committed:
Ensure postgres is installed for tests

moshthepitt committed:
Fix variable name

moshthepitt committed:
Fix broken test

moshthepitt committed:
Add a way to define database connections

- Remove pgpass var
- Add var for SQL to set up db

moshthepitt committed:
Build pgpass using db connections

moshthepitt committed:
Add task to run setup SQL

moshthepitt committed:
Add db connection test vars

moshthepitt committed:
Fix lint issue

moshthepitt committed:
Add database_migrations_sqitch_plans var

moshthepitt committed:
Add tasks to run sqitch migrations

moshthepitt committed:
Add support for different engines

moshthepitt committed:
Set defaults for registry, extra_args, and engine

moshthepitt committed:
Remove database_migrations_system_user_create_home

moshthepitt committed:
Add database_migrations_sql_jobs

moshthepitt committed:
Add mechanism to set cron jobs for sql scripts

moshthepitt committed:
Cleanup

moshthepitt committed:
Add more test vars

moshthepitt committed:
Refactor how cron vars are set

Have them set from cron.yml for better efficiency.

moshthepitt committed:
Fix lint issue

moshthepitt committed:
Add more CRON options

moshthepitt committed:
Add database_migrations_log_path

moshthepitt committed:
Add logs for cron jobs

moshthepitt committed:
Update test inventory

moshthepitt committed:
Uncomment code

moshthepitt committed:
Remove no-needed conditions

moshthepitt committed:
Move to database_migrations_create_pgpass vars/main.yml

moshthepitt committed:
Add comments to explain variables

moshthepitt committed:
Add documentation

moshthepitt committed:
Remove unused handlers

moshthepitt committed:
Update test vars to use Reveal

moshthepitt committed:
Minor clean up

moshthepitt committed:
Verify sqitch migrations

moshthepitt committed:
Add test for pgpass

moshthepitt committed:
Add test for cron

moshthepitt committed:
Add more tests

- Test that setup SQL was run successfully
- Test that periodic SQL is run okay
- Assert that periodic SQL is output to log file

moshthepitt committed:
Add database_migrations_sys_name

moshthepitt committed:
Add task to confugre logrotate

moshthepitt committed:
Skip idempotence test

moshthepitt committed:
Enhance documentation

ukanga commented on README.md:
```diff
@@ -1,35 +1,259 @@
+### database_migrations_system_user
 
-Including an example of how to use your role (for instance, with variables passed in as parameters) is always nice for users too:
+This is the user which is used while running the role.  The default is "nomad" - you know, because they migrate :)
```
I wonder, should we name the repository and role `ansible-nomad`? Or `ansible-sqitch-migrations`?

moshthepitt replied:
I'll try `ansible-sqitch-migrations` :) - ansible-nomad is tempting but I think the generally accepted way to name ansible roles is in a way that describes what they do

moshthepitt replied:
Done

ukanga commented on tasks/setup.yml:
```diff
@@ -0,0 +1,62 @@
+  with_items:
+    - "{{ database_migrations_log_path }}"
+
+- name: Copy pgpass credentials
+  template:
+    src: "pgpass.j2"
+    dest: "{{ database_migrations_home }}/.pgpass"
+    owner: "{{ database_migrations_system_user }}"
+    group: "{{ database_migrations_system_group }}"
+    mode: "600"
+  when:
+    - database_migrations_db_connections.values() | length > 0
+    - database_migrations_create_pgpass | bool
```
Why is this required given we have `db/setup.yml` below?

moshthepitt replied:
@ukanga I was thinking that one may want to run/rerun this role and not touch any existing .pgpass giles

ukanga commented on tasks/clone.yml:
```diff
@@ -0,0 +1,13 @@
+---
+- name: Clone {{ repo.url }}
+  include_role:
+    name: adamyala.ansible-role-clone-repo
```
Is there a benefit we get when we use this role instead of the default Ansible [Git module](https://docs.ansible.com/ansible/latest/modules/git_module.html)?

moshthepitt replied:
It provides a way to have the git stuff done in a more packaged and re-usable way.  Very similar to what [onaio.ansible-git-clone](https://github.com/onaio/ansible-git-clone) is for.  I am suing this instead of `onaio.ansible-git-clone` because right now onaio.ansible-git-clone is broken for ssh agent forwarding.

ukanga commented on meta/main.yml:
```diff
@@ -1,16 +1,27 @@
+dependencies:
+  - role: onaio.sqitch
+    become: true
+    become_user: "root"
```
The default value is `root`, I would exclude this unless the user is someone else other than `root`.

moshthepitt replied:
100% will remove this.

moshthepitt committed:
Fix Github actions test

- Use molecule-action v2.3.4
- Specify tox.ini file
- Use python 3.7 for test
- Ensure python 3.7 is installed okay

moshthepitt committed:
Remove tox file

moshthepitt committed:
Add tox file

moshthepitt committed:
Ignore tox directory while linting

moshthepitt committed:
Specify python3.8 for tests

moshthepitt committed:
Ignore tox directory

moshthepitt committed:
Git clone with https instead of ssh

moshthepitt committed:
Don't specify ssh-agent

moshthepitt committed:
Add tox to python requirements

moshthepitt committed:
Remove not-necessary become

moshthepitt committed:
Change name to ansible-sqitch-migrations

gstuder-ona commented on README.md:
````diff
@@ -1,35 +1,259 @@
+An object containing SQL queries that should be executed periodically.
+
+```yml
+sqitch_migrations_sql_jobs:
````
yay!

IMO we should move to as db-centered an approach as we can, so we can define mat views alongside their refresh intervals (something like a system.cron table or pg_cron) vs requiring redeploy, but we can then just use this to run a SP or something to check dynamically what needs to happen in the next time slice.

moshthepitt replied:
Yes I agree.  We should push as much as possible to the DB itself.  However I think when we are using Amazon's RDS our hands are a little tied because they don't allow pg_cron.  Anyway, I think we are generally on the same page and should watch out that this is not abused.

gstuder-ona commented on README.md:
```diff
@@ -1,35 +1,259 @@
+        - sqitch_migrations_sqitch_plans:
+            db_admin:
+              - directory: "{{ sqitch_migrations_home }}/data-solutions/examples/sqitch_test"
+            db_user:
```
This one is a little scary to me - if we start assembling a lot of micro-migrations together in playbooks vs data-solutions repo, we won't be able to test stuff very well without a lot of infrastructure.

Do we want to push more of this down to data solutions - you can definitely use subdirectories and symlinks, for example, for shared migration piece.  Alternately copy-paste, if each project is really a separate fork at the time-of-creation, could make sense in this case.

EDIT: On second glance I'd say it's more a usage thing, in that we can decide what granularity to break up our scripts but it's still useful here to be able to specify multiple migrations

moshthepitt replied:
I've actually been really torn about how the Reveal migrations are currently structured.  I might change it.  Maybe.  Anyway, I think sqitch was meant to be used more like what you describe here:

> Do we want to push more of this down to data solutions - you can definitely use subdirectories and symlinks, for example, for shared migration piece. Alternately copy-paste, if each project is really a separate fork at the time-of-creation, could make sense in this case.

So I do agree that we should be wary of using this role to orchestrate migrations.  We should let sqitch do that.

Though as noted in your edit we should let it be possible to run multiple migrations e.g. in the case of discover.ona.io we may have migrations for more than one projects living in the same ansible inventory.

gstuder-ona reviewed (commented):
Looks good, more of a question - would it make sense to by-default include ansible-sqitch as a dependency (with a flag to turn that off)?  I know for git-clone, for example, which is similar, it's annoying to always forget to install git first and Stouts.backup doesn't make you install duplicity separately.

moshthepitt commented:
> Looks good, more of a question - would it make sense to by-default include ansible-sqitch as a dependency (with a flag to turn that off)? I know for git-clone, for example, which is similar, it's annoying to always forget to install git first and Stouts.backup doesn't make you install duplicity separately

It is set up as a dependency.  If you don't have sqitch running then this role will install it ready for use with postgres.
